compute_sdk/tests/unit/test_executor.py

```diff
@@ -229,43 +264,98 @@ def test_executor_context_manager(gc_executor):
+
+def test_executor_shutdown_idempotent_with(
+    gc_executor: t.Tuple[Client, Executor], mock_log
+):
     _, gce = gc_executor
-    assertion = AssertionError("Expected that context manager exit is noop")
-    mock_rw = mock.Mock(spec=_ResultWatcher)
-    mock_rw.shutdown.side_effect = (None, assertion)
-    gce._result_watcher = mock_rw
     with gce:
         gce.shutdown()
-        gce._result_watcher = mock_rw
-    gce._result_watcher = None
+        assert "initiating shutdown" in mock_log.debug.call_args_list[0][0][0]
+        assert mock_log.debug.call_count == 3
+    assert mock_log.debug.call_count == 3
+
+
+@pytest.mark.parametrize("wait", (True, False))
+def test_executor_shutdown_wait(
+    wait: bool,
+    mocker: MockerFixture,
+    gc_executor: t.Tuple[Client, Executor],
+    mock_result_watcher: mock.Mock,
+):
+    _, gce = gc_executor
+    gce.task_group_id = uuid.uuid4()
+
+    task_ids = [str(uuid.uuid4()) for _ in range(random.randint(1, 2))]
+    open_futs = {task_id: ComputeFuture(task_id=task_id) for task_id in task_ids}
+    mock_result_watcher.get_open_futures.return_value = open_futs
+    mocker.patch(
+        f"{_MOCK_BASE}_RESULT_WATCHERS", new={gce.task_group_id: mock_result_watcher}
+    )
+    mock_wait = mocker.patch.object(concurrent.futures, "wait")
+
+    gce.shutdown(wait=wait)
+
+    if wait:
+        assert mock_wait.called
+        assert list(mock_wait.call_args[0][0]) == list(open_futs.values())
+    else:
+        assert not mock_wait.called
+
+
+@pytest.mark.parametrize("cancel_futures", (True, False))
+def test_executor_shutdown_cancel_futures(
+    cancel_futures: bool, gc_executor: t.Tuple[Client, Executor]
+):
+    gcc, gce = gc_executor
+    gce.endpoint_id = uuid.uuid4()
+    gce.task_group_id = uuid.uuid4()
+
+    gcc.register_function.return_value = str(uuid.uuid4())
+
+    gce._task_submitter = mock.Mock(spec=threading.Thread)
+    gce._task_submitter.join.side_effect = lambda: None
+    gce._task_submitter.is_alive.return_value = True
+
+    futures: list[ComputeFuture] = []
+    num_submits = random.randint(1, 20)
+    for _ in range(num_submits):
+        future = gce.submit(noop)
+        futures.append(future)
+
+    assert gce._tasks_to_send.qsize() == num_submits, "Verify test setup"
+
+    gce.shutdown(cancel_futures=cancel_futures)
+
+    assert gce._tasks_to_send.qsize() == 1  # Only poison pill remains
+    if cancel_futures:
+        assert all(fut.cancelled() for fut in futures)
+    else:
+        assert all(not fut.cancelled() for fut in futures)
+
+    gce._task_submitter.is_alive.return_value = False  # Allow test cleanup
 
 
 @pytest.mark.parametrize("num_submits", (random.randint(1, 1000),))
-def test_executor_stuck_submitter_doesnt_hold_shutdown(
-    gc_executor, num_submits, mock_log
+def test_executor_stuck_submitter_doesnt_hold_shutdown_if_cancel_futures(
+    gc_executor: t.Tuple[Client, Executor], num_submits: int
```

Comment:
This is a subtle change from what we'd originally implemented.  In particular, the choice I had made originally was to assume that if the script was shutting down, then any not-yet-sent tasks hadn't been created yet and the expectation was to shutdown now.

This now requires that the user specify `cancel_futures` in such a scenario.  A reasonable choice&nbsp;&mdash;&nbsp;arguably API-correct even, if not the most user-intuitive&nbsp;&mdash;&nbsp;but I think we need to call this out in the user-facing documentation as I suspect that interaction will be annoying the uninitiated.  Along the lines of "Why do I have to hit Ctrl+c so many times?!"

Reply:
It is a subtle change, but I think it moves us closer to matching the standard Python `Executor` parent class, whose [documentation](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor.shutdown) states the following: 

> If cancel_futures is True, this method will cancel all pending futures that the executor has not started running. Any futures that are completed or running won’t be cancelled, regardless of the value of cancel_futures.

We currently cancel all futures that have not been registered with the web service (i.e., pending) even if `cancel_futures` is `False`. This diverts from the parent's `.shutdown()` method and, I think, is less user-intuitive.

On a related note, this PR implicitly considers any future that has been registered with the web service as "running."  I think this is a reasonable stance because we cannot know the point at which the remote machine has allocated resources for a specific task, and we currently have no way to stop a task once it has been registered with the web services. I can see an argument that `cancel_futures=True` should cancel all futures, but that would further divert from the parent class (IMO) and would have no stop the tasks from ultimately running.

Reply:
I added some clarifying verbiage on the `.shutdown()` docstring.

Reply:
Comments left in dev, but docstring additional is definitely a useful call-out.

Reply:
@rjmello Here's my read on this: 
``` python

with Executor(...) as ex:
    futures = [ex.submit(foo) for i in range(10)]
    sleep(<SOME_SECONDS>)

print([fu.done() for fu in futures]) # inconsistent states dependent on whether the tasks had enough time to submit
```

We currently have inconsistent behavior based on whether there's a delay that allows the tasks to move from the pending queue to being submitted to the service and being tracked by the result watcher when the context closes. This is bad. I think behaving with `cancel_futures=False` as the default where we always try to move tasks to the service is what is consistent with the executor interface.

Reply:
Just to clarify, I'm all for this change. While arguably the change is "subtle", the behavior that a user will see will be more consistent. I've run into issues where the context closes with some futures submitted and this fixes that.